ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageCompactor.java

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.cloud;
+
+import com.codahale.metrics.Timer;
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.replication.PartitionInfo;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class CloudStorageCompactor implements Runnable {
+  private static final Logger logger = LoggerFactory.getLogger(CloudStorageCompactor.class);
+  private final CloudDestination cloudDestination;
+  private final Map<PartitionId, PartitionInfo> partitionMap;
+  private final VcrMetrics vcrMetrics;
+  private final boolean testMode;
+
+  /**
+   * Public constructor.
+   * @param cloudDestination the cloud destination to use.
+   * @param partitionMap the map whose keys are the partitions to compact.
+   * @param vcrMetrics the metrics to update.
+   * @param testMode if true, the compaction methods only perform the query but do not actually remove the blobs.
+   */
+  public CloudStorageCompactor(CloudDestination cloudDestination, Map<PartitionId, PartitionInfo> partitionMap,
+      VcrMetrics vcrMetrics, boolean testMode) {
+    this.cloudDestination = cloudDestination;
+    this.partitionMap = partitionMap;
+    this.vcrMetrics = vcrMetrics;
+    this.testMode = testMode;
+  }
+
+  @Override
+  public void run() {
+    Timer.Context docTimer = vcrMetrics.blobCompactionTime.time();
+    try {
+      compactPartitions();
+    } finally {
+      docTimer.stop();
+    }
```

Comment:
One gotcha i've found with ScheduledExecutorService is that, if an execution throws an exception, future executions of the runnable will stop: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleAtFixedRate(java.lang.Runnable,%20long,%20long,%20java.util.concurrent.TimeUnit)

I would recommend catching any runtimeexceptions (Maybe throwables too?) and logging here.

Reply:
Will do.